core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -717,64 +698,34 @@ class KafkaController(val config: KafkaConfig,
    *
    * @throws IllegalStateException if a partition is not in `partitionsBeingReassigned`
    */
-  private def maybeTriggerPartitionReassignment(topicPartitions: Set[TopicPartition]): Map[TopicPartition, ApiError] = {
-    val reassignmentResults: mutable.Map[TopicPartition, ApiError] = mutable.Map.empty
-    val partitionsToBeRemovedFromReassignment = scala.collection.mutable.Set.empty[TopicPartition]
+  private def maybeTriggerPartitionReassignment(reassignments: Map[TopicPartition, ReassignmentContext]): Map[TopicPartition, ApiError] = {
+    reassignments.map { case (tp, reassignmentContext) =>
+      val topic = tp.topic
 
-    topicPartitions.foreach { tp =>
-      if (topicDeletionManager.isTopicQueuedUpForDeletion(tp.topic)) {
+      val apiError = if (topicDeletionManager.isTopicQueuedUpForDeletion(topic)) {
         info(s"Skipping reassignment of $tp since the topic is currently being deleted")
-        partitionsToBeRemovedFromReassignment.add(tp)
-        reassignmentResults.put(tp, new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION, "The partition does not exist."))
+        new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION, "The partition does not exist.")
       } else {
-        val reassignedPartitionContext = controllerContext.partitionsBeingReassigned.get(tp).getOrElse {
-          throw new IllegalStateException(s"Initiating reassign replicas for partition $tp not present in " +
-            s"partitionsBeingReassigned: ${controllerContext.partitionsBeingReassigned.mkString(", ")}")
-        }
-        val newReplicas = reassignedPartitionContext.newReplicas
-        val topic = tp.topic
         val assignedReplicas = controllerContext.partitionReplicaAssignment(tp)
         if (assignedReplicas.nonEmpty) {
-          if (assignedReplicas == newReplicas) {
-            info(s"Partition $tp to be reassigned is already assigned to replicas " +
```

Comment:
It seemed simpler not to optimize for this case and let it fall through the normal reassignment completion logic. I think the only cost is a round of UpdateMetadata requests. Maybe that's fine considering it's weird to do a no-op reassignment. What do you think?

Reply:
I don't mind. I always prefer simplicity over performance in non-essential paths and this seems like one

Reply:
Well, `moveReassignedPartitionLeaderIfRequired` would bump up the leader epoch and send a LAIR wave too, right?

Reply:
True. I think I still prefer not to optimize for this case since it simplifies the handling. One (debatable) benefit is that we could use a no-op reassignment to force LeaderAndIsr propagation, which would be lighter than a controller move.

Reply:
Agreed